File: wwwroot/js/skybox.js
function SkyboxProgram(vertexShaderURL, fragmentShaderURL) {
	console.log("new skybox program: vsURL=" + vertexShaderURL + " fsURL=" + fragmentShaderURL);
	this.modelList = [];
	
	// Async request for shader program
	var p = this; // don't put 'this' inside the closure below
	fetchProgramFromURL(vertexShaderURL, fragmentShaderURL, function (prog) { skyboxShaderLoaded(p, prog); });
}

function skyboxGetAttr(p, attr) {
	p[attr] = gl.getAttribLocation(p.shaderProgram, attr);
	if (p[attr] < 0) {
		console.log("skybox: failure querying attribute location: " + attr);
	}
}

function skyboxGetUniform(p, uniform) {
	p[uniform] = gl.getUniformLocation(p.shaderProgram, uniform);
	if (p[uniform] < 0) {
		console.log("skybox: failure querying uniform location: " + uniform);
	}
}

function skyboxShaderLoaded(p, prog) {

	if (!('shaderProgram' in prog)) {
		console.log("skybox: shader program load failure");
		return;
	}

	console.log("skybox: shader program loaded");
	p.shaderProgram = prog.shaderProgram;
		
	// save attribute location
	skyboxGetAttr(p, "a_Position");

	// save uniform location
	skyboxGetUniform(p, "u_MV");
	skyboxGetUniform(p, "u_P");
	skyboxGetUniform(p, "u_Skybox");
}

SkyboxProgram.prototype.addModel = function(m) {
	this.modelList.push(m);
}

SkyboxProgram.prototype.drawModels = function() {
	
    gl.useProgram(this.shaderProgram);
    gl.enableVertexAttribArray(this.aPosition);
	
	/*
	var unit = 0;
	gl.activeTexture(gl.TEXTURE0 + unit);
	gl.uniform1i(this.u_Skybox, unit);
	*/

	gl.uniformMatrix4fv(this.u_P, false, neg.pMatrix);
	
	for (var m in this.modelList) {
		this.modelList[m].drawInstances();
	}
	
	// clean up
	gl.bindBuffer(gl.ARRAY_BUFFER, null);
	gl.bindBuffer(gl.ELEMENT_ARRAY_BUFFER, null);
	
    //gl.disableVertexAttribArray(this.aVertexPosition); // needed ??
}

function SkyboxModel(program, URL, reverse, rescale) {
	this.program = program;
	this.URL = URL;
	this.instanceList = [];
	
	this.cubemapTexture = gl.createTexture();
	gl.bindTexture(gl.TEXTURE_CUBE_MAP, this.cubemapTexture);
	gl.texParameteri(gl.TEXTURE_CUBE_MAP, gl.TEXTURE_MAG_FILTER, gl.LINEAR);
	gl.texParameteri(gl.TEXTURE_CUBE_MAP, gl.TEXTURE_MIN_FILTER, gl.LINEAR);
	
	// Async request for buffer data (mesh)
	var m = this; // don't put 'this' inside the closure below
	fetchBufferData(this.URL, function (buf) { skyboxModelBufferDataLoaded(m, buf); }, reverse, rescale);	
}

function skyboxModelBufferDataLoaded(model, buf, reverse) {
	model.buffer = buf;
}

SkyboxModel.prototype.addInstance = function(i) {
	this.instanceList.push(i);
}

SkyboxModel.prototype.drawInstances = function() {

	gl.bindTexture(gl.TEXTURE_CUBE_MAP, this.cubemapTexture);

	for (var i in this.instanceList) {
		this.instanceList[i].draw(this.program);
	}
	
	gl.bindTexture(gl.TEXTURE_CUBE_MAP, null);
}

SkyboxModel.prototype.addCubemapFace = function(face, URL) {
	console.log("add cubemap face: " + URL);
	loadCubemapFace(face, this.cubemapTexture, URL);
}

function SkyboxInstance(model, center, scale) {
	this.model = model;
	this.center = center;
	this.scale = scale;
}

SkyboxInstance.prototype.draw = function(program) {

	var buf = this.model.buffer;

	// view transform
	//mat4.identity(this.viewMatrix);
	//mat4.lookAt([0,0,0], [0,0,-1], [0,1,0], this.viewMatrix);

	var MV = mat4.create(); // model-view

	// 6/7. camera
	mat4.lookAt(neg.eye, neg.center, neg.up, MV);
	
	/*
	// 5. obj translate
    mat4.translate(MV, this.center);
		
	// 1. obj scale
	mat4.scale(MV, [this.scale, this.scale, this.scale]);
	*/
	gl.uniformMatrix4fv(program.u_MV, false, MV);
	
	// vertex coord
    gl.bindBuffer(gl.ARRAY_BUFFER, buf.vertexPositionBuffer);
   	gl.vertexAttribPointer(program.aVertexPosition, buf.vertexPositionBufferItemSize, gl.FLOAT, false, 0, 0);

	// cubemap sampler
	var unit = 0;
	gl.activeTexture(gl.TEXTURE0 + unit);
	gl.bindTexture(gl.TEXTURE_CUBE_MAP, this.model.cubemapTexture);
	gl.uniform1i(program.u_Skybox, unit);
	
	// draw
	gl.bindBuffer(gl.ELEMENT_ARRAY_BUFFER, buf.vertexIndexBuffer);
	gl.drawElements(gl.TRIANGLES, buf.vertexIndexLength, gl.UNSIGNED_SHORT, 0 * buf.vertexIndexBufferItemSize);
}

File: wwwroot/dart/skybox.dart
library skybox;

import 'dart:html';
import 'dart:web_gl';
import 'dart:math' as math;

import 'package:vector_math/vector_math.dart';
import 'package:game_loop/game_loop_html.dart';

import 'shader.dart';
import 'camera.dart';
import 'interpolate.dart';
import 'logg.dart';
import 'texture.dart';

class SkyboxProgram extends ShaderProgram {
  UniformLocation u_Skybox;

  SkyboxProgram(RenderingContext gl) : super(gl, "skyboxShader");

  void getLocations() {
    super.getLocations();

    u_Skybox = gl.getUniformLocation(program, "u_Skybox");
  }

  /*
  void fetch(Map<String,Shader> shaderCache, String vertexShaderURL, fragmentShaderURL) {
    super.fetch(shaderCache, vertexShaderURL, fragmentShaderURL);
  }
  */

  void drawModels(GameLoopHtml gameLoop, Camera cam, Matrix4 pMatrix) {
    if (!shaderReady) {
      return;
    }

    gl.useProgram(program);
    gl.enableVertexAttribArray(a_Position);

    /*
    int unit = 0;
    gl.activeTexture(RenderingContext.TEXTURE0 + textureUnit);
    gl.uniform1i(this.u_Skybox, unit);
    */

    // send perspective projection matrix uniform
    gl.uniformMatrix4fv(this.u_P, false, pMatrix.storage);

    modelList.forEach((m) => m.drawInstances(gameLoop, this, cam));

    // clean up
    gl.bindBuffer(RenderingContext.ARRAY_BUFFER, null);
    gl.bindBuffer(RenderingContext.ELEMENT_ARRAY_BUFFER, null);

    //gl.disableVertexAttribArray(a_Position); // needed ??
  }
}

class SkyboxModel extends Model {
  Texture cubemapTexture;
  int loadedCubemapFaces = 0;
  final int textureUnit = 0;

  double get halfEdge => 1.0;

  SkyboxModel.fromJson(
      RenderingContext gl, String name, String URL, bool reverse)
      : super.fromJson(gl, name, URL, reverse) {
    cubemapTexture = gl.createTexture();
  }

  void addCubemapFace(RenderingContext gl, int face, String URL) {
    ImageElement image = new ImageElement();

    void handleDone(Event e) {

      // bind texture cubemapTexture to unit textureUnit
      //gl.activeTexture(RenderingContext.TEXTURE0 + textureUnit);
      gl.bindTexture(RenderingContext.TEXTURE_CUBE_MAP, cubemapTexture);

      gl.texParameteri(RenderingContext.TEXTURE_CUBE_MAP,
          RenderingContext.TEXTURE_MAG_FILTER, RenderingContext.NEAREST);
      gl.texParameteri(RenderingContext.TEXTURE_CUBE_MAP,
          RenderingContext.TEXTURE_MIN_FILTER, RenderingContext.NEAREST);

      gl.texImage2DImage(face, 0, RenderingContext.RGBA, RenderingContext.RGBA,
          RenderingContext.UNSIGNED_BYTE, image);

      gl.texParameteri(RenderingContext.TEXTURE_CUBE_MAP,
          RenderingContext.TEXTURE_WRAP_S, RenderingContext.CLAMP_TO_EDGE);
      gl.texParameteri(RenderingContext.TEXTURE_CUBE_MAP,
          RenderingContext.TEXTURE_WRAP_T, RenderingContext.CLAMP_TO_EDGE);

      //anisotropic_filtering_enable(gl);

      gl.bindTexture(RenderingContext.TEXTURE_CUBE_MAP, null);

      ++loadedCubemapFaces;
    }

    void handleError(Event e) {
      err("addCubemapFace: handleError: failure loading image from URL: $URL: $e");
    }

    image
      ..onLoad.listen(handleDone)
      ..onError.listen(handleError)
      ..src = URL;
  }

  void drawInstances(GameLoopHtml gameLoop, ShaderProgram program, Camera cam) {
    if (!modelReady || loadedCubemapFaces < 6) {
      return;
    }

    RenderingContext gl = program.gl;

    // bind textureUnit to cubemapTexture
    //gl.activeTexture(RenderingContext.TEXTURE0 + textureUnit);
    gl.bindTexture(RenderingContext.TEXTURE_CUBE_MAP, cubemapTexture);

    // set sampler to use texture assigned to textureUnit
    //gl.uniform1i((program as SkyboxProgram).u_Skybox, textureUnit);
    gl.uniform1i((program as SkyboxProgram).u_Skybox, defaultTextureUnit);

    instanceList.forEach((Instance i) => i.draw(gameLoop, program, cam));

    gl.bindTexture(RenderingContext.TEXTURE_CUBE_MAP, null);
  }
}

class SkyboxInstance extends Instance {

  // demo animate fields:
  final double degreesPerSec = 20.0;
  bool demoAnimate;
  double _oldAngle = 0.0;
  double _angle = 0.0;

  double get halfEdge => scale * (model as SkyboxModel).halfEdge;

  SkyboxInstance(
      String id, Model model, Vector3 center, double scale, this.demoAnimate)
      : super(id, model, center, scale);

  void update(GameLoopHtml gameLoop) {
    if (demoAnimate) {
      _oldAngle = _angle;
      _angle = gameLoop.gameTime * this.degreesPerSec % 360.0;
    }
  }

  void draw(GameLoopHtml gameLoop, ShaderProgram prog, Camera cam) {
    double rescale;

    if (demoAnimate) {
      double deg = interpolateDegree(
          _angle, _oldAngle, gameLoop.renderInterpolationFactor);
      double rad = deg * math.PI / 180.0;
      double size = 15 * math.sin(rad).abs() + 1;
      rescale = scale * size;
    } else {
      rescale = scale;
    }

    RenderingContext gl = prog.gl;

    uploadModelView(gl, prog.u_MV, cam, rescale); // set up MV matrix

    gl.bindBuffer(RenderingContext.ARRAY_BUFFER, model.vertexPositionBuffer);
    gl.vertexAttribPointer(prog.a_Position, model.vertexPositionBufferItemSize,
        RenderingContext.FLOAT, false, 0, 0);

    gl.bindBuffer(
        RenderingContext.ELEMENT_ARRAY_BUFFER, model.vertexIndexBuffer);

    model.pieceList.forEach((piece) {

      //gl.bindTexture(RenderingContext.TEXTURE_CUBE_MAP, (model as SkyboxModel).cubemapTexture);

      gl.drawElements(RenderingContext.TRIANGLES, piece.vertexIndexLength,
          RenderingContext.UNSIGNED_SHORT,
          piece.vertexIndexOffset * model.vertexIndexBufferItemSize);
    });
  }
}
